pkg/proxy/conntrack/cleanup.go

```diff
@@ -59,57 +64,126 @@ func deleteStaleServiceConntrackEntries(ct Interface, svcPortMap proxy.ServicePo
-		}
+		filters = append(filters, filterForIP(svcIP, v1.ProtocolUDP))
 	}
 	klog.V(4).InfoS("Deleting conntrack stale entries for services", "nodePorts", conntrackCleanupServiceNodePorts.UnsortedList())
 	for _, nodePort := range conntrackCleanupServiceNodePorts.UnsortedList() {
-		err := ct.ClearEntriesForPort(nodePort, isIPv6, v1.ProtocolUDP)
-		if err != nil {
-			klog.ErrorS(err, "Failed to clear udp conntrack", "nodePort", nodePort)
-		}
+		filters = append(filters, filterForPort(nodePort, v1.ProtocolUDP))
+	}
+
+	if err := ct.ClearEntries(getUnixIPFamily(isIPv6), filters...); err != nil {
+		klog.ErrorS(err, "Failed to delete stale service connections")
 	}
 }
 
 // deleteStaleEndpointConntrackEntries takes care of flushing stale conntrack entries related
 // to UDP endpoints. After a UDP endpoint is removed we must flush any conntrack entries
 // for it so that if the same client keeps sending, the packets will get routed to a new endpoint.
 func deleteStaleEndpointConntrackEntries(ct Interface, svcPortMap proxy.ServicePortMap, endpointsUpdateResult proxy.UpdateEndpointsMapResult) {
+	var filters []netlink.CustomConntrackFilter
+	isIPv6 := false
 	for _, epSvcPair := range endpointsUpdateResult.DeletedUDPEndpoints {
 		if svcInfo, ok := svcPortMap[epSvcPair.ServicePortName]; ok {
+			isIPv6 = netutils.IsIPv6(svcInfo.ClusterIP())
 			endpointIP := proxyutil.IPPart(epSvcPair.Endpoint)
 			nodePort := svcInfo.NodePort()
-			var err error
 			if nodePort != 0 {
-				err = ct.ClearEntriesForPortNAT(endpointIP, nodePort, v1.ProtocolUDP)
-				if err != nil {
-					klog.ErrorS(err, "Failed to delete nodeport-related endpoint connections", "servicePortName", epSvcPair.ServicePortName)
-				}
-			}
-			err = ct.ClearEntriesForNAT(svcInfo.ClusterIP().String(), endpointIP, v1.ProtocolUDP)
-			if err != nil {
-				klog.ErrorS(err, "Failed to delete endpoint connections", "servicePortName", epSvcPair.ServicePortName)
+				filters = append(filters, filterForPortNAT(endpointIP, nodePort, v1.ProtocolUDP))
+
 			}
+			filters = append(filters, filterForNAT(svcInfo.ClusterIP().String(), endpointIP, v1.ProtocolUDP))
 			for _, extIP := range svcInfo.ExternalIPs() {
-				err := ct.ClearEntriesForNAT(extIP.String(), endpointIP, v1.ProtocolUDP)
-				if err != nil {
-					klog.ErrorS(err, "Failed to delete endpoint connections for externalIP", "servicePortName", epSvcPair.ServicePortName, "externalIP", extIP)
-				}
+				filters = append(filters, filterForNAT(extIP.String(), endpointIP, v1.ProtocolUDP))
 			}
 			for _, lbIP := range svcInfo.LoadBalancerVIPs() {
-				err := ct.ClearEntriesForNAT(lbIP.String(), endpointIP, v1.ProtocolUDP)
-				if err != nil {
-					klog.ErrorS(err, "Failed to delete endpoint connections for LoadBalancerIP", "servicePortName", epSvcPair.ServicePortName, "loadBalancerIP", lbIP)
-				}
+				filters = append(filters, filterForNAT(lbIP.String(), endpointIP, v1.ProtocolUDP))
```

Comment:
TODO: stop flipping back and forth between `net.IP` and `string`. (Note that `filterForNAT` is just going to immediately re-parse this string back to `net.IP`.) But this will require changes to `pkg/proxy`.